### PR TITLE
Give `IGate` a no-op definition

### LINKED ISSUE #4010: Gate.control() method rasies the error 'Cannot make controlled gate with id instruction'
### Information

- **Qiskit Terra version**:
0.12.0
- **Python version**:
Python 3.7.3
- **Operating system**:
WIN10

### What is the current behavior?
Converting circuit into gate using the `QuantumCircuit.to_gate` property, and then applying the `Gate.control` method raises the error 'Cannot make controlled gate with id instruction'. This actually means that you cannot create a control gate if you have an id instruction in your circuit. However, when the circuit with no `IdGate` is used, the Gate.control work just fine.

### Steps to reproduce the problem

```
CIRCF_2.iso(B2, [0, 1, 2, 3],[]) # works fine 
GATEB2=transpile(CIRCF_2, basis_gates=['u1', 'u2', 'u3', 'cx', 'id']).to_gate() # works fine
CGA=GATEB2.control(1) # crashes with error 'Cannot make controlled gate with id instruction'
```

### What is the expected behavior?
The Identity gate `IdGate` should not compromise the control instruction.

### LINKED ISSUE #7399: Generating controlled gate from QuantumCircuit results in AttributeError when circuit contains at least one explicit identity operator
### Environment

- **Qiskit Terra version**: 0.19.1 (Also tested this in 0.18.0 with same outcome)
- **Python version**: 3.9.7
- **Operating system**: CentOS (Also tested this on MacOS 12.0.1 with same outcome) 

### What is happening?

When you take a `QuantumCircuit` where one of its constituent gates is explicitly the identity and you attempt to generate a controlled gate from this unitary using the `control()` method, Qiskit raises the error:

`AttributeError: 'NoneType' object has no attribute 'global_phase'`

### How can we reproduce the issue?

Examples of code that reproduce this error:

```
from qiskit.opflow.operator_globals import X, Y, Z, I

pauli_str_op = Z^Z^Z^I

pauli_str_op.to_circuit().control()
```

```
from qiskit import Aer, QuantumCircuit, QuantumRegister, AncillaRegister, ClassicalRegister
from qiskit.utils.algorithm_globals import algorithm_globals
from qiskit.utils.backend_utils import is_aer_provider
from qiskit.opflow.operator_globals import X, Y, Z, I


pauli_str_op = Z^Z^Z^I

qreg = QuantumRegister(4, 'q')
circ = QuantumCircuit(qreg)
circ.z(qreg[0])
circ.z(qreg[1])
circ.z(qreg[2])
circ.id(qreg[3])

circ.control()
```

This results in the following Traceback:

```
Traceback (most recent call last):
  File "/Users/joelbierman/Desktop/Qiskit_Projects/control_gate_error.py", line 6, in <module>
    pauli_str_op.to_circuit().control()
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/quantumcircuit.py", line 554, in control
    controlled_gate = gate.control(num_ctrl_qubits, label, ctrl_state)
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/gate.py", line 119, in control
    return add_control(self, num_ctrl_qubits, label, ctrl_state)
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/add_control.py", line 59, in add_control
    cgate = control(operation, num_ctrl_qubits=num_ctrl_qubits, label=label, ctrl_state=ctrl_state)
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/add_control.py", line 113, in control
    unrolled_gate = _unroll_gate(operation, basis_gates=basis)
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/circuit/add_control.py", line 268, in _unroll_gate
    opqc = dag_to_circuit(unroller.run(dag))
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/transpiler/passes/basis/unroller.py", line 113, in run
    unrolled_dag = self.run(decomposition)  # recursively unroll ops
  File "/Users/joelbierman/opt/anaconda3/envs/Qiskit_0280_pyscf2_ARM64/lib/python3.9/site-packages/qiskit/transpiler/passes/basis/unroller.py", line 76, in run
    raise QiskitError(
qiskit.exceptions.QiskitError: "Error decomposing node of instruction 'id': 'NoneType' object has no attribute 'global_phase'. Unable to define instruction 'id' in the given basis."
```

### What should happen?

Controlled unitary of `QuantumCircuit` is generated, regardless of what unitary gates make up the circuit.

### Any suggestions?

This seems to be related to how Qiskit handles generating controlled unitaries from circuits that explicitly contain at least one identity operator. This of course does not happen if the identity implicitly acts on one of the qubits since this would, by definition, break the `control()` method for all circuits. For example, the following code does not raise this error:

```
from qiskit import Aer, QuantumCircuit, QuantumRegister, AncillaRegister, ClassicalRegister
from qiskit.utils.algorithm_globals import algorithm_globals
from qiskit.utils.backend_utils import is_aer_provider
from qiskit.opflow.operator_globals import X, Y, Z, I


pauli_str_op = Z^Z^Z^I

qreg = QuantumRegister(4, 'q')
circ = QuantumCircuit(qreg)
circ.z(qreg[0])
circ.z(qreg[1])
circ.z(qreg[2])
# nothing acts on qubit 3: the identity is implicit.
circ.control()

```

### LINKED ISSUE #9485: `transpile` function with `basis_gates=['u', 'cx']` fails with any circuit with an `id` gate
### Environment

- **Qiskit Terra version**: 0.22.3
- **Python version**: 3.9.15

### What is happening?

Suppose I am given an arbitrary quantum circuit and I want to transpile it by using only the gate $U(\theta, \phi, \lambda)$ (single-qubit rotation) and the gate $CX$ (creating entanglement), knowing that they form a universal set of gates for quantum computation.
In Qiskit this should be possible calling the `transpile` function and passing the argument `basis_gates=['u', 'cx']`. This works fine in almost any case but fails for every circuit with an identity `id` gate (or more than one).

### How can we reproduce the issue?

```
from qiskit import QuantumCircuit, transpile

qc = QuantumCircuit(1)
qc.id(0)
tqc = transpile(qc, basis_gates=['u', 'cx'])
```

`QiskitError: "Cannot unroll the circuit to the given basis, ['u', 'cx']. Instruction id not found in equivalence library and no rule found to expand."`

### What should happen?

I would expect the transpiler not to raise any error and to complete the circuit transpilation correctly.

### Any suggestions?

To solve this problem, it should be enough to add $I = U(0,0,0)$ to the Qiskit equivalence library, so that the transpiler runs successfully and returns the circuit
ugate

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This allows it to work with all passes that decompose operations using
`Instruction.definition`, even if there is no particular special
handling for it.

### Details and comments

Fix #7399.
Fix #4010.
Fix #9485.

### CODE REVIEW

**Comment by jakelishman** [comment]
I am marginally concerned that those newly uncovered lines indicate a problem somewhere in our testing logic; they're meant to throw an error, but we didn't have any test failures before, and we don't now.  Seems like somewhere might be overzealously catching exceptions.

**Review by kdk — COMMENTED** [review]
Similar to #7146 , I think the approach here is inconsistent with how Qiskit currently defines `IGate`.
https://github.com/Qiskit/qiskit-terra/blob/309bfb173e946f441243f292a7655203419efde9/qiskit/circuit/library/standard_gates/i.py#L23

This behavior is deprecated but not yet removed from the IBMQ backends at which point we can `IGate` to it's canonical definition. Until then, perhaps we can raise a more meaningful error message when someone tries to unroll, decompose, or control an `IGate`.

File: qiskit/circuit/library/standard_gates/i.py [hunk 1]
```diff
@@ -48,6 +48,11 @@ def __init__(self, label: Optional[str] = None):
         """Create new Identity gate."""
         super().__init__("id", 1, [], label=label)
 
+    def _define(self) -> None:
```
Comment (kdk) [inline qiskit/circuit/library/standard_gates/i.py:51]: Can you update the docstring above to remove the section about `IGate` being defined as a delay and note that it is now explicitly a nop?

File: releasenotes/notes/fix-igate-definition-7385cff02e663867.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    :meth:`.QuantumCircuit.control`, would crash if an identity operation
+    represented by an :class:`.IGate` was present in a :class:`.QuantumCircuit`.
+    For example, the following code was previously an error, but will now function
+    correctly::
+
+        from qiskit import QuantumCircuit
+
+        qc = QuantumCircuit(1)
+        qc.i(0)
+        controlled = qc.control()
```
Comment (kdk) [inline releasenotes/notes/fix-igate-definition-7385cff02e663867.yaml:13]: Can you also include a change note that the definition of `IGate` has changed?

**Review by kdk — COMMENTED** [review]
This will be great to be able to finally reconcile. A few documentation comments but otherwise LGTM.

File: qiskit/circuit/library/standard_gates/i.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/circuit/library/standard_gates/i.py:51]: Oh yeah, thanks for the reminder: done in 95ec70c5d.

File: releasenotes/notes/fix-igate-definition-7385cff02e663867.yaml (hunk 2 above)
Comment (jakelishman) [reply releasenotes/notes/fix-igate-definition-7385cff02e663867.yaml:13]: Done in 95ec70c5d.

**Review by mtreinish — COMMENTED** [review]
This LGTM, and I think this will mostly supersede #7146 so I think we should just do this for 0.23.0 and we can revisit whether #7146 is needed at all or not after the release

**Comment by kdk** [comment]
After a discussion with @taalexander and @dtmcclure , I'd like to wait a bit on this. It looks like there's a desire to keep the `'id'` instruction on IBM hardware as a way to have an idle instruction with a well-defined duration. If that's the case, there might be a better definition for `IGate` than an empty circuit (like `U2Gate(0, 0)`).

**Comment by jakelishman** [comment]
But the transpiler will also remove `U2Gate(0, 0)` in its optimisation passes, which I think is actually a bit worse than this setup, which allows the IBM backends to insert their custom translation steps to convert `IGate` to whatever they'd like, and that dodges this definition (and since `id` is in their basis sets, we wouldn't even _use_ this conversion).

**Comment by jakelishman** [comment]
I really don't like the proprietary IBM hardware preventing the backend-agnostic Qiskit from using the natural mathematical name `id` to refer to the identity operation, since it co-opts it to mean something different, especially when there's plenty of ways through Qiskit to let the hardware do whatever it wants, without affecting us.

Not having this be a true identity operation keeps causing problems at higher-levels, where algorithms and applications (via `SparsePauliOp`, for example), can't output a natural identity operation, but have to do it implicitly only.  We can work-around them (and repeatedly have to), but it would be much easier if those guys didn't have to.

**Comment by kdk** [comment]
Agree that the result post-optimization of most common uses cases with the proposals on the table will look very similar in practice but I would like, if possible, to find a pathway forward that is consistent with how `'id'` has historically been defined in Qiskit and has been understood by users.

This might look like defining `IGate` to be an identity/nop single-qubit instruction as a `UGate(0,0,0)` (apologies if this seems overly pedantic compared to the existing empty definition, but I want to be sure the unoptimized behavior can be represented by a do-nothing instruction on hardware). Then we can note that it may have a backend-dependent (but potentially non-zero) duration, that it can be expected to be optimized away by the transpiler at all optimization_levels > 0. Hopefully, that will resolve the conflict of `'id'` being opaque for uses cases that want to rely on optimization and preserves the definition of `'id'` as an instruction which can have an implementation on hardware.

File: qiskit/circuit/library/standard_gates/i.py [hunk 3]
```diff
@@ -48,6 +48,11 @@ def __init__(self, label: Optional[str] = None):
...
+        from qiskit.circuit import QuantumCircuit  # pylint: disable=cyclic-import
+
+        self.definition = QuantumCircuit(1, name=self.name)
```
Comment (kdk) [inline qiskit/circuit/library/standard_gates/i.py:54]:
```suggestion
        self.definition = QuantumCircuit(1, name=self.name)
        self.definition.u(0, 0, 0, 0)
```

File: qiskit/circuit/library/standard_gates/i.py [hunk 4]
```diff
@@ -20,12 +20,12 @@
 class IGate(Gate):
     r"""Identity gate.
 
-    Identity gate corresponds to a single-qubit gate wait cycle,
-    and should not be optimized or unrolled (it is an opaque gate).
+    Logically this gate is a quantum no-op and the transpiler will largely treat it as such, but
+    some older hardware may choose to interpret it as a short-hand for the shortest-possible delay
+    cycle.
```
Comment (kdk) [inline qiskit/circuit/library/standard_gates/i.py:25]:
```suggestion
    This gate is a quantum no-op and represents an identity channel on a single qubit.
    It will not have a logical effect on the qubit state and will likely be optimized away by the compiler at all optimization levels greater than zero.
    Hardware backends may implement this instruction with a potentially non-zero idle duration, for example a delay of length equal to that of a single-qubit gate, but this behavior will depend on the selected backend.
```

File: qiskit/circuit/library/standard_gates/i.py (hunk 3 above)
Comment (jakelishman) [reply qiskit/circuit/library/standard_gates/i.py:54]: I don't think this achieves anything; if a backend wants to assign further meaning to the operation then they can put it in their `basis_gates` in which case this definition will never get used, and if it _isn't_ in `basis_gates`, then unitary optimisation will remove the gate with-or-without this extra gate in the definition.

File: releasenotes/notes/fix-igate-definition-7385cff02e663867.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+upgrade:
+  - |
+    Qiskit Terra will now generally treat the identity gate :class:`.IGate` (with corresponding
+    circuit method :meth:`.QuantumCircuit.i`) as a quantum no-op, and more aggressively remove it
+    during calls to :func:`.transpile`.  This is a change to previous behaviour that supported
+    legacy IBM Quantum systems where the ``id`` instruction was a short-hand for the shortest
+    possible delay cycle.
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply releasenotes/notes/fix-igate-definition-7385cff02e663867.yaml]: I did something mostly similar to this in d6e385457 - I found the "single qubit pulse length delay" a bit hard to parse, so I changed it slightly in a way that'd be good if you checked it.

File: qiskit/circuit/library/standard_gates/i.py (hunk 4 above)
Comment (jakelishman) [reply qiskit/circuit/library/standard_gates/i.py]: Done (give or take) in d6e385457.

**Review by mtreinish — APPROVED** [review]
I think now is the time to do this, I can't think of a use case that will be broken by this. For IBM backends with their custom behavior for `id` that will still work because `id` is in the basis gates for those backend and they can treat it special (there is also the custom translation plugin the ibm provider has to to translate id to single qubit gate duration delays). For other backends if they support id that will continue to work as is, but for backends that don't support `id` this fixes transpilation for those backends.

**Comment by mtreinish** [comment]
I'm fine with this or #7146, the specific approach doesn't matter to me. What I think is critical to fix is that `transpile()` with an `id` gate can be successfully run targeting a backend that doesn't have a native `id` instruction. That being an issue users have to deal with has been sitting around for too long.

**Comment by kdk** [comment]
Closing in favor of https://github.com/Qiskit/qiskit/pull/7146 .
